internal/decision/engine.go

```diff
@@ -185,27 +232,43 @@ func newEngine(bs bstore.Blockstore, bstoreWorkerCount int, peerTagger PeerTagge
 	e := &Engine{
 		ledgerMap:                       make(map[peer.ID]*ledger),
 		scoreLedger:                     scoreLedger,
-		bsm:                             newBlockstoreManager(bs, bstoreWorkerCount),
+		bsm:                             newBlockstoreManager(ctx, bs, bstoreWorkerCount, pendingBlocksGauge, activeBlocksGauge),
 		peerTagger:                      peerTagger,
 		outbox:                          make(chan (<-chan *Envelope), outboxChanBuffer),
 		workSignal:                      make(chan struct{}, 1),
 		ticker:                          time.NewTicker(time.Millisecond * 100),
 		maxBlockSizeReplaceHasWithBlock: maxReplaceSize,
-		taskWorkerCount:                 taskWorkerCount,
+		taskWorkerCount:                 engineTaskWorkerCount,
 		sendDontHaves:                   true,
 		self:                            self,
 		peerLedger:                      newPeerLedger(),
+		pendingGauge:                    pendingEngineGauge,
+		activeGauge:                     activeEngineGauge,
 	}
 	e.tagQueued = fmt.Sprintf(tagFormat, "queued", uuid.New().String())
 	e.tagUseful = fmt.Sprintf(tagFormat, "useful", uuid.New().String())
 	e.peerRequestQueue = peertaskqueue.New(
 		peertaskqueue.OnPeerAddedHook(e.onPeerAdded),
 		peertaskqueue.OnPeerRemovedHook(e.onPeerRemoved),
 		peertaskqueue.TaskMerger(newTaskMerger()),
-		peertaskqueue.IgnoreFreezing(true))
+		peertaskqueue.IgnoreFreezing(true),
+		peertaskqueue.MaxOutstandingWorkPerPeer(maxOutstandingBytesPerPeer))
 	return e
 }
 
+func (e *Engine) updateMetrics() {
+	if e.metricUpdateCounter%100 == 0 {
+		stats := e.peerRequestQueue.Stats()
+		e.activeGauge.Set(float64(stats.NumActive))
+		e.pendingGauge.Set(float64(stats.NumPending))
+	}
+	e.metricUpdateCounter++
+}
```

Comment:
This is a data race since `updateMetrics` is called inside the engine worker goroutines. It seems like there's a bit of an impedance issue here since we're polling (i.e. pull) the peertaskqueue to get stats to set gauges (i.e. push).

Would putting function options into peertaskqueue that say that a given instance should use the given gauges (i.e. still named with bitswap) be reasonable, or is that too much time spent on a potentially important pathway?

> e.metricUpdateCounter%100 == 0

Could we go a long time without seeing any metrics updates here in a way that might confuse us?

Reply:
Fixed the race.

> Would putting function options into peertaskqueue that say that a given instance should use the given gauges (i.e. still named with bitswap) be reasonable, or is that too much time spent on a potentially important pathway?

I don't fully understand your concern here with calling PTQ.Stats(). Tasks take orders of magnitude more time than taking a lock to read an integer.

> Could we go a long time without seeing any metrics updates here in a way that might confuse us?

I don't think so. There's always a value in the counter, the prometheus server polls the counter on regular intervals, so there's always a number on the display. The "%" logic just determines when the counter is updated.

Reply:
> I don't fully understand your concern here with calling PTQ.Stats(). Tasks take orders of magnitude more time than taking a lock to read an integer.

It's likely fine, but without measuring performance it's hard to tell. Locking/Unlocking themselves take some time and that lock is synchronized across all of the active workers which is a little concerning. It's something we may want to track during the next RC phase of go-ipfs to be sure there aren't regressions though.

> the prometheus server polls the counter on regular intervals

This is what makes me feel like something is off. The gauges are push metrics (i.e. you push updates to them), but it looks like prometheus supports pull metrics too (e.g. they're used for collecting the TCP metrics https://github.com/libp2p/go-tcp-transport/blob/c2f0ef4161b9726862f8a39d120107fe37a48c7e/metrics.go#L33). So we're pulling metrics to put into the push metrics which is pretty wonky.

Maybe the issue here is that go-libp2p-interface-metrics (or at least the implementation in https://github.com/ipfs/go-metrics-prometheus) only seems to support the pull metrics. I'm not sure the history with those repos so we can probably call this good enough, track this during the next RC, and file an issue to go through what to do about pull metrics.